### DISCUSSION TITLE
Does this model support streaming?

### DISCUSSION THREAD

**Discussion by AnkushMalaker** [moonshine-ai/moonshine#18] [Q&A]
If so, how is the performance compared to recently released [https://github.com/revdotcom/reverb](reverb) or [https://github.com/ufal/whisper_streaming](whisper_streaming)?
And can you provide an example snippet or script?

If not, I'm a little confused, acc to the blog post, a huge focus was to run this on edge devices like raspberry pis and microcontrollers where streaming becomes important.
Thanks for your awesome work!

**Comment by guynich** [comment]
@AnkushMalaker 
Moonshine is a batch model.  We provide the open-source demo [`live_captions.py`](https://github.com/usefulsensors/moonshine/blob/main/moonshine/demo/live_captions.py) as an example of how to implement repeated model inferences with live speech.  We have not quantified performance of our demo against other methods such as your links.  I believe those may also use batch models at the core too, such as Whisper, with possibly similar wrapper approaches to enable streaming-like operation.  Thanks for your interest.

**Reply by nirbhaysinghnarang** [reply]
Are you guys using the same approach for the Torre device? 
Also, distil-whisper-small runs just fine on my iPhone (with an `onnx` runtime) - do you have plans to release a model that compares in terms of feature size? If I recall correctly, moonshine only offers models comparable to the tiny and base versions of whisper - is that correct?

**Reply by keveman** [reply]
Yes, we are running Moonshine on Torre. Currently we have `tiny` and `base` versions of Moonshine that have equal or better accuracies compared to Whisper's `tiny.en` and `base.en` models respectively, but run multiple times faster for short audio sequences. We do have `onnx` versions available : https://github.com/usefulsensors/moonshine/blob/main/moonshine/demo/onnx_standalone.py

**Reply by nirbhaysinghnarang** [reply]
I understand that. However, what I was getting at is more along the lines of asking if you have any plans of supporting the `small.en` equivalent?

**Reply by nirbhaysinghnarang** [reply]
Interesting that you are running Moonshine on Torre. The demo indicates real-time transcription - are you following the algorithm outlined in `demos/live_captions.py` i.e. using VAD to detect utterances? If so, I would suggest doing something a little different. The fine folks at `sherpa-onnx` have a really solid 2-pass implementation where they use a pretty inaccurate online, streaming model with endpoint detection to not only detect utterances (which are then passed into an offline model like moonshine or whisper) but also give interim updates.

**Reply by keveman** [reply]
The 2 pass method used in `sherpa-onnx` is indeed interesting. However, in our trials, using a 20M `zipformer` as the first pass caused pretty bad first pass output, which is not surprising. It eventually got all fixed up correctly by the second pass (whisper base in our trials). In our implementation, we use `moonshine` itself for both passes and `silero-vad` to detect pauses. Empirically, the quality looked better because we are using the same high accuracy model for both passes. However, I confess the quality of output from our streaming implementation is not as good as it can be and we are constantly improving it.

**Reply by AnkushMalaker** [reply]
Its hard to use if streaming performance is not good. For any application where you'd want STT on edge (home assistant/voice assistant), it depends on good streaming performance. 
I suppose this is still good for replacing whisper for large scale ASR. Helps the pockets. Thanks for your work.

**Comment by csukuangfj** [comment]
For those of you who want to try the two-pass method with moonshine models in sherpa-onnx, please download pre-built Android APKs from
https://k2-fsa.github.io/sherpa/onnx/android/apk-2pass.html

![Screenshot 2024-10-31 at 15 16 37](https://github.com/user-attachments/assets/defe5f7f-a009-4d6f-aefb-336b1236104a)


----

APKs for silero_vad + moonshine models are at
https://k2-fsa.github.io/sherpa/onnx/vad/apk-asr.html

![Screenshot 2024-10-31 at 15 18 01](https://github.com/user-attachments/assets/535d2f3e-50a6-416f-8b87-b46f32886c55)

**Reply by nirbhaysinghnarang** [reply]
Is the code for these demos open-sourced?

**Reply by csukuangfj** [reply]
Yes, everything from Next-gen Kaldi is open-sourced, including sherpa-onnx.

please see
https://github.com/k2-fsa/sherpa-onnx/tree/master/android

**Reply by csukuangfj** [reply]
Hint: Doc is at https://k2-fsa.github.io/sherpa/onnx
